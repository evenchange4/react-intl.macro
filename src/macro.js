// @flow
/* eslint no-console: 0 */
import path from 'path';
import { createMacro } from 'babel-plugin-macros';
import * as R from 'ramda';
import {
  type Babel,
  type BabelPluginPass,
  type BabelPath,
} from 'babel-flow-types';
import { importReactIntl, writeFileSync, getMessages } from './utils/index';
// import printAST from 'ast-pretty-print';
// console.log(printAST(referencePath.parentPath));

function reactIntlMacro({
  references,
  state,
  state: {
    file: {
      opts: { filename },
    },
  },
  babel: { types: t },
}: {
  references: { defineMessages: Array<BabelPath> },
  state: BabelPluginPass,
  babel: Babel,
}): void {
  const { defineMessages = [] } = references;
  const { MESSAGE_DIR } = process.env;

  // Note: add react-intl on the top of file if it have not been imported.
  const calleeName = defineMessages[0].parentPath.get('callee').node.name;
  importReactIntl(state, t, calleeName);

  // Note: Output side-effect
  if (MESSAGE_DIR) {
    const result = defineMessages.map(getMessages);
    const sourceRelativedDir = path.relative(process.cwd(), filename);
    const inputFilename = filename;
    const outputFilename = path
      .join(process.cwd(), MESSAGE_DIR, sourceRelativedDir)
      .replace(/(js|jsx)$/g, 'json');

    console.log(
      `${path.relative(process.cwd(), inputFilename)} -> ${path.relative(
        process.cwd(),
        outputFilename,
      )}`,
    );
    writeFileSync(
      {
        filename: outputFilename,
        data: R.flatten(result),
      },
      state,
    );
  }
}

export default createMacro(reactIntlMacro);
