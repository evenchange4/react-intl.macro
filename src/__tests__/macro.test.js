// @flow

import pluginTester from 'babel-plugin-tester';
import plugin from 'babel-plugin-macros';

expect.addSnapshotSerializer({
  print(val) {
    return val.replace(/..\/macro/, 'react-intl.macro');
  },
  test(val) {
    return typeof val === 'string';
  },
});

pluginTester({
  plugin,
  snapshot: true,
  babelOptions: {
    filename: __filename,
    parserOpts: { plugins: ['jsx'] },
  },
  tests: {
    '[defineMessages] one X one': {
      error: false,
      code: `
        import { defineMessages } from '../macro';

        const messages = defineMessages({
          'Component.greet': {
            id: 'Component.greet',
            defaultMessage: 'Hello, {name}!',
            description: 'Greeting to welcome the user to the app',
          },
        });

        export default messages;
      `,
    },
    '[defineMessages] one X two': {
      error: false,
      code: `
        import { defineMessages } from '../macro';

        const messages = defineMessages({
          'Component.greet': {
            id: 'Component.greet',
            defaultMessage: 'Hello, {name}!',
            description: 'Greeting to welcome the user to the app',
          },
          'Component.greet2': {
            id: 'Component.greet2',
            defaultMessage: 'Hello, {name}!',
            description: 'Greeting to welcome the user to the app',
          },
        });

        export default messages;
      `,
    },
    '[defineMessages] two X one': {
      error: false,
      code: `
        import { defineMessages } from '../macro';

        export const message1 = defineMessages({
          'Component.greet2': {
            id: 'Component.greet2',
            defaultMessage: 'Hello, {name}!',
            description: 'Greeting to welcome the user to the app',
          },
        });

        export const message2 = defineMessages({
          'Component.greet': {
            id: 'Component.greet',
            defaultMessage: 'Hello, {name}!',
            description: 'Greeting to welcome the user to the app',
          },
        });
      `,
    },
    '[defineMessages] it should only import react-intl once': {
      error: false,
      code: `
        import { injectIntl } from 'react-intl';
        import { defineMessages } from '../macro';

        const messages = defineMessages({
          'Component.greet': {
            id: 'Component.greet',
            defaultMessage: 'Hello, {name}!',
            description: 'Greeting to welcome the user to the app',
          },
        });

        export default messages;
      `,
    },
    '[defineMessages] callee alias': {
      error: false,
      code: `
        import { injectIntl } from 'react-intl';
        import { defineMessages as def } from '../macro';

        const messages = def({
          'Component.greet': {
            id: 'Component.greet',
            defaultMessage: 'Hello, {name}!',
            description: 'Greeting to welcome the user to the app',
          },
        });

        export default messages;
      `,
    },
    '[FormattedMessage] one X one': {
      error: false,
      code: `
        import * as React from 'react';
        import { FormattedMessage } from '../macro';

        export default class Foo extends React.Component {
          render() {
            return (
              <FormattedMessage
                id="Foo.hello"
                defaultMessage='Hello, {name}!'
                description='Greeting to welcome the user to the app'
              />
            );
          }
        }
      `,
    },
    '[FormattedMessage] one X two': {
      error: false,
      code: `
        import * as React from 'react';
        import { FormattedMessage } from '../macro';

        export default class Foo extends React.Component {
          render() {
            return (
              <React.Fragment>
                <FormattedMessage
                  id="Foo.hello1"
                  defaultMessage='Hello, {name}!'
                  description='Greeting to welcome the user to the app'
                />
                <FormattedMessage
                  id="Foo.hello2"
                  defaultMessage='Hello, {name}!'
                  description='Greeting to welcome the user to the app'
                />
              </React.Fragment>
            );
          }
        }
      `,
    },
    '[FormattedMessage] two X one': {
      error: false,
      code: `
        import * as React from 'react';
        import { FormattedMessage } from '../macro';

        class Foo extends React.Component {
          render() {
            return (
              <FormattedMessage
                id="Foo.hello"
                defaultMessage='Hello, {name}!'
                description='Greeting to welcome the user to the app'
              />
            );
          }
        }

        const Component = () => (
          <FormattedMessage
            id="Component.hello"
            defaultMessage='Hello, {name}!'
            description='Greeting to welcome the user to the app'
          />
        )
      `,
    },
    '[FormattedHTMLMessage]': {
      error: false,
      code: `
        import * as React from 'react';
        import { FormattedHTMLMessage } from '../macro';

        export default class Foo extends React.Component {
          render() {
            return (
              <FormattedHTMLMessage
                id="FormattedHTMLMessage.hello"
                defaultMessage='<div>Hello, {name}!</div>'
                description='Greeting to welcome the user to the app'
              />
            );
          }
        }
      `,
    },
    '[defineMessages, FormattedMessage, FormattedHTMLMessage]': {
      error: false,
      code: `
        import * as React from 'react';
        import { defineMessages, FormattedMessage, FormattedHTMLMessage } from '../macro';

        const messages = defineMessages({
          'Component.greet': {
            id: 'defineMessages.greet',
            defaultMessage: 'Hello, {name}!',
            description: 'Greeting to welcome the user to the app',
          },
        });

        class Foo extends React.Component {
          render() {
            return (
              <React.Fragment>
                <FormattedMessage
                  id="Foo.hello1"
                  defaultMessage='Hello, {name}!'
                  description='Greeting to welcome the user to the app'
                />
                <FormattedMessage
                  id="Foo.hello2"
                  defaultMessage='Hello, {name}!'
                  description='Greeting to welcome the user to the app'
                />
              </React.Fragment>
            );
          }
        }

        const Component = () => (
          <FormattedHTMLMessage
            id="FormattedHTMLMessage.hello"
            defaultMessage='<div>Hello, {name}!</div>'
            description='Greeting to welcome the user to the app'
          />
        )
      `,
    },
  },
});
