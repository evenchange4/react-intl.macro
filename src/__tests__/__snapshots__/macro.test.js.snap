// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros [FormattedHTMLMessage]: [FormattedHTMLMessage] 1`] = `

import * as React from 'react';
import { FormattedHTMLMessage } from 'react-intl.macro';

export default class Foo extends React.Component {
  render() {
    return (
      <FormattedHTMLMessage
        id="FormattedHTMLMessage.hello"
        defaultMessage='<div>Hello, {name}!</div>'
        description='Greeting to welcome the user to the app'
      />
    );
  }
}

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedHTMLMessage } from "react-intl";
import * as React from 'react';
export default class Foo extends React.Component {
  render() {
    return <FormattedHTMLMessage id="FormattedHTMLMessage.hello" defaultMessage='<div>Hello, {name}!</div>' description='Greeting to welcome the user to the app' />;
  }

}
[{
  "id": "FormattedHTMLMessage.hello",
  "defaultMessage": "<div>Hello, {name}!</div>",
  "description": "Greeting to welcome the user to the app"
}]

`;

exports[`macros [FormattedMessage] multiline with TemplateLiteral issue#38: [FormattedMessage] multiline with TemplateLiteral issue#38 1`] = `

import * as React from 'react';
import { FormattedMessage } from 'react-intl.macro';

export default class Foo extends React.Component {
  render() {
    return (
      <FormattedMessage
        id="Foo.hello"
        defaultMessage={\`Hello {name}, you have {unreadCount, number} {unreadCount, plural,
          one {message}
          other {messages}
        }\`}
        values={{name: <b>{name}</b>, unreadCount}}
      />
    );
  }
}

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedMessage } from "react-intl";
import * as React from 'react';
export default class Foo extends React.Component {
  render() {
    return <FormattedMessage id="Foo.hello" defaultMessage={\`Hello {name}, you have {unreadCount, number} {unreadCount, plural,
          one {message}
          other {messages}
        }\`} values={{
      name: <b>{name}</b>,
      unreadCount
    }} />;
  }

}
[{
  "id": "Foo.hello",
  "defaultMessage": "Hello {name}, you have {unreadCount, number} {unreadCount,plural,one{message} other{messages}}"
}]

`;

exports[`macros [FormattedMessage] one X one: [FormattedMessage] one X one 1`] = `

import * as React from 'react';
import { FormattedMessage } from 'react-intl.macro';

export default class Foo extends React.Component {
  render() {
    return (
      <FormattedMessage
        id="Foo.hello"
        defaultMessage='Hello, {name}!'
        description='Greeting to welcome the user to the app'
      />
    );
  }
}

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedMessage } from "react-intl";
import * as React from 'react';
export default class Foo extends React.Component {
  render() {
    return <FormattedMessage id="Foo.hello" defaultMessage='Hello, {name}!' description='Greeting to welcome the user to the app' />;
  }

}
[{
  "id": "Foo.hello",
  "defaultMessage": "Hello, {name}!",
  "description": "Greeting to welcome the user to the app"
}]

`;

exports[`macros [FormattedMessage] one X two: [FormattedMessage] one X two 1`] = `

import * as React from 'react';
import { FormattedMessage } from 'react-intl.macro';

export default class Foo extends React.Component {
  render() {
    return (
      <React.Fragment>
        <FormattedMessage
          id={1}
          defaultMessage='Hello, {name}!'
          description='Greeting to welcome the user to the app'
        />
        <FormattedMessage
          id={2}
          defaultMessage='Hello, {name}!'
          description='Greeting to welcome the user to the app'
        />
      </React.Fragment>
    );
  }
}

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedMessage } from "react-intl";
import * as React from 'react';
export default class Foo extends React.Component {
  render() {
    return <React.Fragment>
        <FormattedMessage id={1} defaultMessage='Hello, {name}!' description='Greeting to welcome the user to the app' />
        <FormattedMessage id={2} defaultMessage='Hello, {name}!' description='Greeting to welcome the user to the app' />
      </React.Fragment>;
  }

}
[{
  "id": 1,
  "defaultMessage": "Hello, {name}!",
  "description": "Greeting to welcome the user to the app"
}, {
  "id": 2,
  "defaultMessage": "Hello, {name}!",
  "description": "Greeting to welcome the user to the app"
}]

`;

exports[`macros [FormattedMessage] should evaluate with expression issue#38: [FormattedMessage] should evaluate with expression issue#38 1`] = `

import * as React from 'react';
import { FormattedMessage } from 'react-intl.macro';

export default class Foo extends React.Component {
  render() {
    return (
      <FormattedMessage
        id={0 + 1}
        defaultMessage={\`\${1 + 2} Hello {name}, you have {unreadCount, number} {unreadCount, plural,
          one {message}
          other {messages}
        }\`}
        description={\`description \${1+ 1}\`}
        values={{name: <b>{name}</b>, unreadCount}}
      />
    );
  }
}

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedMessage } from "react-intl";
import * as React from 'react';
export default class Foo extends React.Component {
  render() {
    return <FormattedMessage id={0 + 1} defaultMessage={\`\${1 + 2} Hello {name}, you have {unreadCount, number} {unreadCount, plural,
          one {message}
          other {messages}
        }\`} description={\`description \${1 + 1}\`} values={{
      name: <b>{name}</b>,
      unreadCount
    }} />;
  }

}
[{
  "id": 1,
  "defaultMessage": "3 Hello {name}, you have {unreadCount, number} {unreadCount,plural,one{message} other{messages}}",
  "description": "description 2"
}]

`;

exports[`macros [FormattedMessage] two X one: [FormattedMessage] two X one 1`] = `

import * as React from 'react';
import { FormattedMessage } from 'react-intl.macro';

class Foo extends React.Component {
  render() {
    return (
      <FormattedMessage
        id="Foo.hello"
        defaultMessage='Hello, {name}!'
        description='Greeting to welcome the user to the app'
      />
    );
  }
}

const Component = () => (
  <FormattedMessage
    id="Component.hello"
    defaultMessage='Hello, {name}!'
    description='Greeting to welcome the user to the app'
  />
)

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedMessage } from "react-intl";
import * as React from 'react';

class Foo extends React.Component {
  render() {
    return <FormattedMessage id="Foo.hello" defaultMessage='Hello, {name}!' description='Greeting to welcome the user to the app' />;
  }

}

const Component = () => <FormattedMessage id="Component.hello" defaultMessage='Hello, {name}!' description='Greeting to welcome the user to the app' />;

[{
  "id": "Foo.hello",
  "defaultMessage": "Hello, {name}!",
  "description": "Greeting to welcome the user to the app"
}, {
  "id": "Component.hello",
  "defaultMessage": "Hello, {name}!",
  "description": "Greeting to welcome the user to the app"
}]

`;

exports[`macros [defineMessages, FormattedMessage, FormattedHTMLMessage]: [defineMessages, FormattedMessage, FormattedHTMLMessage] 1`] = `

import * as React from 'react';
import { defineMessages, FormattedMessage, FormattedHTMLMessage } from 'react-intl.macro';

const messages = defineMessages({
  'Component.greet': {
    id: 'defineMessages.greet',
    defaultMessage: 'Hello, {name}!',
    description: 'Greeting to welcome the user to the app',
  },
});

class Foo extends React.Component {
  render() {
    return (
      <React.Fragment>
        <FormattedMessage
          id="Foo.hello1"
          defaultMessage='Hello, {name}!'
          description='Greeting to welcome the user to the app'
        />
        <FormattedMessage
          id="Foo.hello2"
          defaultMessage='Hello, {name}!'
          description='Greeting to welcome the user to the app'
        />
      </React.Fragment>
    );
  }
}

const Component = () => (
  <FormattedHTMLMessage
    id="FormattedHTMLMessage.hello"
    defaultMessage='<div>Hello, {name}!</div>'
    description='Greeting to welcome the user to the app'
  />
)

      ↓ ↓ ↓ ↓ ↓ ↓

import { defineMessages, FormattedMessage, FormattedHTMLMessage } from "react-intl";
import * as React from 'react';
const messages = defineMessages({
  'Component.greet': {
    id: 'defineMessages.greet',
    defaultMessage: 'Hello, {name}!',
    description: 'Greeting to welcome the user to the app'
  }
});

class Foo extends React.Component {
  render() {
    return <React.Fragment>
        <FormattedMessage id="Foo.hello1" defaultMessage='Hello, {name}!' description='Greeting to welcome the user to the app' />
        <FormattedMessage id="Foo.hello2" defaultMessage='Hello, {name}!' description='Greeting to welcome the user to the app' />
      </React.Fragment>;
  }

}

const Component = () => <FormattedHTMLMessage id="FormattedHTMLMessage.hello" defaultMessage='<div>Hello, {name}!</div>' description='Greeting to welcome the user to the app' />;

[{
  "id": "defineMessages.greet",
  "defaultMessage": "Hello, {name}!",
  "description": "Greeting to welcome the user to the app"
}, {
  "id": "Foo.hello1",
  "defaultMessage": "Hello, {name}!",
  "description": "Greeting to welcome the user to the app"
}, {
  "id": "Foo.hello2",
  "defaultMessage": "Hello, {name}!",
  "description": "Greeting to welcome the user to the app"
}, {
  "id": "FormattedHTMLMessage.hello",
  "defaultMessage": "<div>Hello, {name}!</div>",
  "description": "Greeting to welcome the user to the app"
}]

`;

exports[`macros [defineMessages] callee alias: [defineMessages] callee alias 1`] = `

import { injectIntl } from 'react-intl';
import { defineMessages as def } from 'react-intl.macro';

const messages = def({
  'Component.greet': {
    id: 'Component.greet',
    defaultMessage: 'Hello, {name}!',
    description: 'Greeting to welcome the user to the app',
  },
});

export default messages;

      ↓ ↓ ↓ ↓ ↓ ↓

import { injectIntl, defineMessages as def } from 'react-intl';
const messages = def({
  'Component.greet': {
    id: 'Component.greet',
    defaultMessage: 'Hello, {name}!',
    description: 'Greeting to welcome the user to the app'
  }
});
export default messages;
[{
  "id": "Component.greet",
  "defaultMessage": "Hello, {name}!",
  "description": "Greeting to welcome the user to the app"
}]

`;

exports[`macros [defineMessages] it should only import react-intl once: [defineMessages] it should only import react-intl once 1`] = `

import { injectIntl } from 'react-intl';
import { defineMessages } from 'react-intl.macro';

const messages = defineMessages({
  'Component.greet': {
    id: 'Component.greet',
    defaultMessage: 'Hello, {name}!',
    description: 'Greeting to welcome the user to the app',
  },
});

export default messages;

      ↓ ↓ ↓ ↓ ↓ ↓

import { injectIntl, defineMessages } from 'react-intl';
const messages = defineMessages({
  'Component.greet': {
    id: 'Component.greet',
    defaultMessage: 'Hello, {name}!',
    description: 'Greeting to welcome the user to the app'
  }
});
export default messages;
[{
  "id": "Component.greet",
  "defaultMessage": "Hello, {name}!",
  "description": "Greeting to welcome the user to the app"
}]

`;

exports[`macros [defineMessages] multiline with TemplateLiteral issue#38: [defineMessages] multiline with TemplateLiteral issue#38 1`] = `

import { defineMessages } from 'react-intl.macro';

const messages = defineMessages({
  'Component.greet': {
    id: 'Component.greet',
    defaultMessage: \`Hello {name}, you have {unreadCount, number} {unreadCount, plural,
      one {message}
      other {messages}
    }\`,
    description: \`Greeting to welcome the user to the app\`,
  },
});

export default messages;

      ↓ ↓ ↓ ↓ ↓ ↓

import { defineMessages } from "react-intl";
const messages = defineMessages({
  'Component.greet': {
    id: 'Component.greet',
    defaultMessage: \`Hello {name}, you have {unreadCount, number} {unreadCount, plural,
      one {message}
      other {messages}
    }\`,
    description: \`Greeting to welcome the user to the app\`
  }
});
export default messages;
[{
  "id": "Component.greet",
  "defaultMessage": "Hello {name}, you have {unreadCount, number} {unreadCount,plural,one{message} other{messages}}",
  "description": "Greeting to welcome the user to the app"
}]

`;

exports[`macros [defineMessages] one X one: [defineMessages] one X one 1`] = `

import { defineMessages } from 'react-intl.macro';

const messages = defineMessages({
  'Component.greet': {
    id: 'Component.greet',
    defaultMessage: 'Hello, {name}!',
    description: 'Greeting to welcome the user to the app',
  },
});

export default messages;

      ↓ ↓ ↓ ↓ ↓ ↓

import { defineMessages } from "react-intl";
const messages = defineMessages({
  'Component.greet': {
    id: 'Component.greet',
    defaultMessage: 'Hello, {name}!',
    description: 'Greeting to welcome the user to the app'
  }
});
export default messages;
[{
  "id": "Component.greet",
  "defaultMessage": "Hello, {name}!",
  "description": "Greeting to welcome the user to the app"
}]

`;

exports[`macros [defineMessages] one X two: [defineMessages] one X two 1`] = `

import { defineMessages } from 'react-intl.macro';

const messages = defineMessages({
  'Component.greet': {
    id: 1,
    defaultMessage: 'Hello, {name}!',
    description: 'Greeting to welcome the user to the app',
  },
  'Component.greet2': {
    id: 2,
    defaultMessage: 'Hello, {name}!',
    description: 'Greeting to welcome the user to the app',
  },
});

export default messages;

      ↓ ↓ ↓ ↓ ↓ ↓

import { defineMessages } from "react-intl";
const messages = defineMessages({
  'Component.greet': {
    id: 1,
    defaultMessage: 'Hello, {name}!',
    description: 'Greeting to welcome the user to the app'
  },
  'Component.greet2': {
    id: 2,
    defaultMessage: 'Hello, {name}!',
    description: 'Greeting to welcome the user to the app'
  }
});
export default messages;
[{
  "id": 1,
  "defaultMessage": "Hello, {name}!",
  "description": "Greeting to welcome the user to the app"
}, {
  "id": 2,
  "defaultMessage": "Hello, {name}!",
  "description": "Greeting to welcome the user to the app"
}]

`;

exports[`macros [defineMessages] should evaluate with expression issue#38: [defineMessages] should evaluate with expression issue#38 1`] = `

import { defineMessages } from 'react-intl.macro';

const messages = defineMessages({
  'Component.greet': {
    id: 0 + 1,
    defaultMessage: \`\${1 + 2} Hello {name}, you have {unreadCount, number} {unreadCount, plural,
      one {message}
      other {messages}
    }\`,
    description: \`Greeting to welcome the user to the app \${1+ 1}\`,
  },
});

export default messages;

      ↓ ↓ ↓ ↓ ↓ ↓

import { defineMessages } from "react-intl";
const messages = defineMessages({
  'Component.greet': {
    id: 0 + 1,
    defaultMessage: \`\${1 + 2} Hello {name}, you have {unreadCount, number} {unreadCount, plural,
      one {message}
      other {messages}
    }\`,
    description: \`Greeting to welcome the user to the app \${1 + 1}\`
  }
});
export default messages;
[{
  "id": 1,
  "defaultMessage": "3 Hello {name}, you have {unreadCount, number} {unreadCount,plural,one{message} other{messages}}",
  "description": "Greeting to welcome the user to the app 2"
}]

`;

exports[`macros [defineMessages] two X one: [defineMessages] two X one 1`] = `

import { defineMessages } from 'react-intl.macro';

export const message1 = defineMessages({
  'Component.greet2': {
    id: 'Component.greet2',
    defaultMessage: 'Hello, {name}!',
    description: 'Greeting to welcome the user to the app',
  },
});

export const message2 = defineMessages({
  'Component.greet': {
    id: 'Component.greet',
    defaultMessage: 'Hello, {name}!',
    description: 'Greeting to welcome the user to the app',
  },
});

      ↓ ↓ ↓ ↓ ↓ ↓

import { defineMessages } from "react-intl";
export const message1 = defineMessages({
  'Component.greet2': {
    id: 'Component.greet2',
    defaultMessage: 'Hello, {name}!',
    description: 'Greeting to welcome the user to the app'
  }
});
export const message2 = defineMessages({
  'Component.greet': {
    id: 'Component.greet',
    defaultMessage: 'Hello, {name}!',
    description: 'Greeting to welcome the user to the app'
  }
});
[{
  "id": "Component.greet2",
  "defaultMessage": "Hello, {name}!",
  "description": "Greeting to welcome the user to the app"
}, {
  "id": "Component.greet",
  "defaultMessage": "Hello, {name}!",
  "description": "Greeting to welcome the user to the app"
}]

`;
